### PR TITLE
Chunked data smoke and boundary cases tests.

### PR DESCRIPTION
**Pull request by grixa**

### CODE REVIEW

File: Bricks/net/http/impl/server.h [hunk 1]
```diff
@@ -478,10 +478,9 @@ class GenericHTTPServerConnection final : public HTTPResponder {
  public:
   // The only constructor parses HTTP headers coming from the socket
   // in the constructor of `message_(connection_)`.
-  GenericHTTPServerConnection(
-      Connection&& c,
-      const typename HTTP_REQUEST_DATA::ConstructionParams& params = typename HTTP_REQUEST_DATA::ConstructionParams())
-      : connection_(std::move(c)), message_(connection_, params) {}
+  template <typename... ARGS>
```
Comment (dkorolev) [inline Bricks/net/http/impl/server.h]: Is this change necessary?

File: Bricks/net/http/test.cc [hunk 2]
```diff
@@ -425,6 +425,175 @@ TEST(PosixHTTPServerTest, ChunkedLargeBodyManyChunks) {
   t.join();
 }
 
+TEST(PosixHTTPServerTest, ChunkedSmoke) {
+  auto EchoServerThreadEntry = [](Socket s) {
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: `const`?

File: Bricks/net/http/test.cc [hunk 3]
```diff
@@ -425,6 +425,175 @@ TEST(PosixHTTPServerTest, ChunkedLargeBodyManyChunks) {
...
+    const auto offset = body.length() - length;
+    for (size_t chunks = length; chunks; chunks = chunks * 2 / 3) {
+      std::string chunked_body;
+      for (size_t i = 0; i < chunks; ++i) {
+        const size_t start = offset + length * i / chunks;
+        const size_t end = offset + length * (i + 1) / chunks;
+        chunked_body += current::strings::Printf("%X\r\n", end - start) + body.substr(start, end - start);
+      }
+
+      thread t(EchoServerThreadEntry, Socket(FLAGS_net_http_test_port));
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: Is there a `using std::thread` somewhere above? Непорядок.

Comment (grixa) [inline Bricks/net/http/test.cc]: Yes, there are three usings above, see cb594ee8. Do you suggest removing them?

Comment (mzhurovich) [inline Bricks/net/http/test.cc]: Yes, let's remove them and use `std::sthread`.

File: Bricks/net/http/impl/server.h (hunk 1 above)
Comment (grixa) [inline Bricks/net/http/impl/server.h]: Well, alternatively we can add two arguments (intial_buffer_size and buffer_growth_k) with same default values as they are in the GenericHTTPRequestData constructor.

File: Bricks/net/http/net_http.vcxproj [hunk 4]
```diff
@@ -87,7 +87,7 @@
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
     <ClCompile>
-      <PreprocessorDefinitions>WIN32;NBRICKS_DEBUG_NET;_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions>WIN32;NCURRENT_BRICKS_DEBUG_NET;_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (dkorolev) [inline Bricks/net/http/net_http.vcxproj]: :+1:

File: Bricks/net/http/test.cc [hunk 5]
```diff
@@ -502,65 +502,65 @@ TEST(PosixHTTPServerTest, ChunkedBoundaryCases) {
...
       "resize the buffer 400 -> 800\n",
       "process a 393 bytes long chunk\n",
       "memmove 1 bytes from offset 398 to the beginning\n",
-      "read 798 bytes (buffer offset 1)\n",
+      "read 798 bytes while requested 798 (buffer offset 1)\n",
       "resize the buffer 800 -> 1600\n",
       "process a 792 bytes long chunk\n",
       "memmove 2 bytes from offset 797 to the beginning\n",
-      "read 1 bytes (buffer offset 2)\n"
+      "read 1 bytes while requested 1597 (buffer offset 2)\n"
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: Am I the only one finding these { 1, 2 } offsets weird?

File: Bricks/net/http/test.cc [hunk 6]
```diff
@@ -573,8 +573,10 @@ TEST(PosixHTTPServerTest, ChunkedBoundaryCases) {
     connection.BlockingWrite(headers + headers_paddings[i] + "\r\n\r\n" + chunked_body, false);
     ExpectToReceive(expect_to_receive, connection);
     t.join();
-    ASSERT_EQ(expected_events_block1[i].size(), current::net::HTTPDataJournal().events.size());
-    for (size_t j = 0; j < expected_events_block1[i].size(); ++j) {
+    EXPECT_EQ(expected_events_block1[i].size(), current::net::HTTPDataJournal().events.size());
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: :+1:

**Comment by dkorolev** [comment]
Thanks @grixa! My only remaining concern are those { 1, 2 } offsets -- looks confusing and imho requires either a comment or a fix.

@mzhurovich ?

File: Bricks/net/http/test.cc (hunk 5 above)
Comment (grixa) [inline Bricks/net/http/test.cc]: But what's wrong with these offsets? Previous line in the journal shows, that unparsed 2 bytes was moved to the beginning of the buffer (it is `0\r`, which can't be parsed until the final `\n` is read). So the next portion of data is read at offset 2..

Comment (dkorolev) [inline Bricks/net/http/test.cc]: И впрямь :)

**Comment by dkorolev** [comment]
Ready to merge! @grixa, green :vertical_traffic_light: ?

**Comment by grixa** [comment]
I think so.

**Comment by dkorolev** [comment]
Merged, CC @sompylasar.

**Comment by sompylasar** [comment]
Have no time to review in full, but this is awesome! Thanks @grixa
